File: apps/central/src/printer/templates.ts
// apps/central/src/printer/templates.ts:1
// Templates 58mm JP-58H — 32 cols (fonte normal) / 48 condensado
// Usando node-thermal-printer (ESC/POS)

export type PedidoPrint = {
  id: string;
  origem: string;
  clienteNome: string;
  telefone: string;
  endereco?: string;
  bairro?: string;
  taxa: number;
  mesa?: number | null;
  comanda?: string | null;
  itens: { nome: string; qtd: number; obs?: string; preco: number; detalhe?: string }[];
  subtotal: number;
  total: number;
  formaPag: string;
  horario: string;
  observacao?: string;
};

function line(char = "-", width = 32) { return char.repeat(width); }
function pad2(n: number) { return n.toFixed(2).padStart(6); }

export function comandaCozinha(p: PedidoPrint): string {
  // Fonte grande, sem preço, com MESA marcada
  const lines: string[] = [];
  lines.push("  *** COZINHA ***");
  lines.push(line("="));
  if (p.mesa) lines.push(`MESA ${String(p.mesa).padStart(2, "0")}  ${p.horario}`);
  else lines.push(`${p.origem.toUpperCase()}  ${p.horario}`);
  lines.push(`${p.clienteNome}  ${p.telefone}`);
  if (p.endereco) lines.push(p.endereco);
  if (p.bairro) lines.push(`Bairro: ${p.bairro}  Taxa R$ ${p.taxa.toFixed(2)}`);
  lines.push(line("-"));
  for (const it of p.itens) {
    lines.push(`${it.qtd}x ${it.nome.toUpperCase()}`);
    if (it.detalhe) lines.push(`   ${it.detalhe}`);
    if (it.obs) lines.push(`   OBS: ${it.obs}`);
  }
  if (p.observacao) lines.push(`OBS GERAL: ${p.observacao}`);
  lines.push(line("="));
  lines.push(`PEDIDO #${p.id.slice(0, 8)}`);
  lines.push("");
  return lines.join("\n");
}

export function ticketCliente(p: PedidoPrint): string {
  const lines: string[] = [];
  lines.push("     PIZZAPDV");
  lines.push("  Sítio / Centro");
  lines.push(line("-"));
  lines.push(`Pedido #${p.id.slice(0, 8)}  ${p.horario}`);
  lines.push(`${p.clienteNome}  ${p.telefone}`);
  if (p.mesa) lines.push(`MESA ${String(p.mesa).padStart(2, "0")}`);
  lines.push(line("-"));
  for (const it of p.itens) {
    const totalItem = (it.preco * it.qtd).toFixed(2);
    lines.push(`${it.qtd}x ${it.nome}`.padEnd(22) + `R$ ${totalItem}`.padStart(10));
    if (it.detalhe) lines.push(`   ${it.detalhe}`);
  }
  lines.push(line("-"));
  lines.push(`Subtotal`.padEnd(22) + `R$ ${pad2(p.subtotal)}`.padStart(10));
  lines.push(`Taxa`.padEnd(22) + `R$ ${pad2(p.taxa)}`.padStart(10));
  lines.push(`TOTAL`.padEnd(22) + `R$ ${pad2(p.total)}`.padStart(10));
  lines.push(line("-"));
  lines.push(`Pagamento: ${p.formaPag}`);
  lines.push("Obrigado pela preferência!");
  lines.push("");
  return lines.join("\n");
}

export function ticketDelivery(p: PedidoPrint): string {
  const lines: string[] = [];
  lines.push("   *** DELIVERY ***");
  lines.push(line("="));
  lines.push(`${p.clienteNome}  ${p.telefone}`);
  lines.push(p.endereco ?? "");
  if (p.bairro) lines.push(`Bairro: ${p.bairro}`);
  lines.push(line("-"));
  for (const it of p.itens) lines.push(`${it.qtd}x ${it.nome} ${it.detalhe ? "(" + it.detalhe + ")" : ""}`);
  lines.push(line("-"));
  lines.push(`TOTAL R$ ${p.total.toFixed(2)}  (${p.formaPag})`);
  lines.push(`Taxa entregador R$ ${p.taxa.toFixed(2)}`);
  lines.push(p.observacao ? `Obs: ${p.observacao}` : "");
  lines.push(line("="));
  return lines.join("\n");
}

File: src/PizzaPDV.Printer/Templates.cs
namespace PizzaPDV.Printer;

public record ItemPrint(string Nome, int Qtd, decimal Preco, string? Detalhe, string? Obs);

public record PedidoPrint(
    string Id,
    string Origem,
    string ClienteNome,
    string Telefone,
    string? Endereco,
    string? Bairro,
    decimal Taxa,
    int? Mesa,
    string? Comanda,
    List<ItemPrint> Itens,
    decimal Subtotal,
    decimal Total,
    string FormaPag,
    string Horario,
    string? Observacao
);

public static class Templates
{
    private static string Line(char c='-', int w=32) => new string(c, w);

    public static string ComandaCozinha(PedidoPrint p)
    {
        var lines = new List<string>();
        lines.Add("  *** COZINHA ***");
        lines.Add(Line('=',32));
        if (p.Mesa != null) lines.Add($"MESA {p.Mesa.Value:D2}  {p.Horario}");
        else lines.Add($"{p.Origem.ToUpper()}  {p.Horario}");
        lines.Add($"{p.ClienteNome}  {p.Telefone}");
        if (!string.IsNullOrEmpty(p.Endereco)) lines.Add(p.Endereco);
        if (!string.IsNullOrEmpty(p.Bairro)) lines.Add($"Bairro: {p.Bairro}  Taxa R$ {p.Taxa:F2}");
        lines.Add(Line('-',32));
        foreach (var it in p.Itens)
        {
            lines.Add($"{it.Qtd}x {it.Nome.ToUpper()}");
            if (!string.IsNullOrEmpty(it.Detalhe)) lines.Add($"   {it.Detalhe}");
            if (!string.IsNullOrEmpty(it.Obs)) lines.Add($"   OBS: {it.Obs}");
        }
        if (!string.IsNullOrEmpty(p.Observacao)) lines.Add($"OBS GERAL: {p.Observacao}");
        lines.Add(Line('=',32));
        lines.Add($"PEDIDO #{p.Id.Substring(0, Math.Min(8,p.Id.Length))}");
        lines.Add("");
        return string.Join("\n", lines);
    }

    public static string TicketCliente(PedidoPrint p)
    {
        var lines = new List<string>();
        lines.Add("     PIZZAPDV");
        lines.Add("  Sitio / Centro");
        lines.Add(Line('-',32));
        lines.Add($"Pedido #{p.Id.Substring(0, Math.Min(8,p.Id.Length))}  {p.Horario}");
        lines.Add($"{p.ClienteNome}  {p.Telefone}");
        if (p.Mesa != null) lines.Add($"MESA {p.Mesa.Value:D2}");
        lines.Add(Line('-',32));
        foreach (var it in p.Itens)
        {
            var totalItem = (it.Preco * it.Qtd).ToString("F2");
            var left = $"{it.Qtd}x {it.Nome}";
            if (left.Length > 22) left = left.Substring(0,22);
            lines.Add(left.PadRight(22) + $"R$ {totalItem}".PadLeft(10));
            if (!string.IsNullOrEmpty(it.Detalhe)) lines.Add($"   {it.Detalhe}");
        }
        lines.Add(Line('-',32));
        lines.Add($"Subtotal".PadRight(22) + $"R$ {p.Subtotal:F2}".PadLeft(10));
        lines.Add($"Taxa".PadRight(22) + $"R$ {p.Taxa:F2}".PadLeft(10));
        lines.Add($"TOTAL".PadRight(22) + $"R$ {p.Total:F2}".PadLeft(10));
        lines.Add(Line('-',32));
        lines.Add($"Pagamento: {p.FormaPag}");
        lines.Add("Obrigado pela preferencia!");
        lines.Add("");
        return string.Join("\n", lines);
    }

    public static string TicketDelivery(PedidoPrint p)
    {
        var lines = new List<string>();
        lines.Add("   *** DELIVERY ***");
        lines.Add(Line('=',32));
        lines.Add($"{p.ClienteNome}  {p.Telefone}");
        lines.Add(p.Endereco ?? "");
        if (!string.IsNullOrEmpty(p.Bairro)) lines.Add($"Bairro: {p.Bairro}");
        lines.Add(Line('-',32));
        foreach (var it in p.Itens) lines.Add($"{it.Qtd}x {it.Nome} {(it.Detalhe!=null? "("+it.Detalhe+")":"")}");
        lines.Add(Line('-',32));
        lines.Add($"TOTAL R$ {p.Total:F2}  ({p.FormaPag})");
        lines.Add($"Taxa entregador R$ {p.Taxa:F2}");
        if (!string.IsNullOrEmpty(p.Observacao)) lines.Add($"Obs: {p.Observacao}");
        lines.Add(Line('=',32));
        return string.Join("\n", lines);
    }
}
